### ISSUE TITLE
Bad accuracy with AVX routine, confirmed with enhanced example MNIST (change last activation function from tanh to softmax)

### ISSUE THREAD

**Issue by beru** [tiny-dnn/tiny-dnn#570]
The issue is revealed in #569

**Comment by pliptor** [comment]
Might 10-5 be a too generous threshold for tolerance in test/test_convolutional_layer.h to verify accuracy of AVX? I can't help much here since I don't have an AVX handy.

**Comment by bhack** [comment]
@pliptor You can emulate with https://software.intel.com/en-us/articles/intel-software-development-emulator

**Comment by pliptor** [comment]
That may be an approach. But I was wondering if there was a reason behind a 10-5 threshold.

**Comment by pliptor** [comment]
@beru I closed #569 . #574 has the freshest result with the HEAD.

**Comment by beru** [comment]
https://github.com/beru/tiny-cnn/commit/bf4fba1c432e43e9aef26f15832b77f1cb1aef30
This change will improve training accuracy, but when I train cifar10, accuracy drops to 1000/10000.
I think that problem is the same one reported in #511

**Comment by beru** [comment]
#410 is probablly related...

**Comment by pliptor** [comment]
Maybe the network death has to do with the "fragility of ReLU"  activation (link below) instead of the usage of AVX? Does the network death go away by replacing ReLU by tan_h? 

http://cs231n.github.io/neural-networks-1/

**Comment by beru** [comment]
Replacing ReLU by tan_h extends life expectancy of the network but the death reamins to appear at a peak of training.

**Comment by pliptor** [comment]
@beru That's interesting. The MNIST net is well documented with respect to differences between the original LeNet-5 (it is almost the same). Do you know how the cifar10 network was specified in the tiny-dnn example? Do you know if there is a standard reference like LeNet-5?

**Comment by beru** [comment]
The same thing happens with internal backend and `max_pooling_layer<relu>`.

@pliptor I don't know about the standard cifar10 network model.
https://www.cs.toronto.edu/~kriz/learning-features-2009-TR.pdf
http://rodrigob.github.io/are_we_there_yet/build/classification_datasets_results.html

**Comment by pliptor** [comment]
@beru Thank you. Time to study again.

**Comment by edgarriba** [comment]
@beru cifar10 is a dataset not a model

**Comment by Randl** [comment]
https://ci.appveyor.com/project/tinydnn/tiny-dnn/build/master-878/job/7ca8udug96863200 May be connected?

**Comment by beru** [comment]
@edgarriba You're totally right about that. The CIFAR-10 is a dataset.... The CIFAR-10 is a dataset... The CIFAR-10 is a dataset...

**Comment by beru** [comment]
@Randl I'm not sure how the erros in the job began to surface. It looks like the previous commit eb82a28
 was fine...

**Comment by pliptor** [comment]
@Randl It's been hard for me to isolate the issues. I have no AVX but I did see a simplified tiny-dnn CIFAR-10 network example (I modified it so I can test a bit on my box without having to wait more than one day) also suddenly die from 68% around epoch 26 or so to 10%.

**Comment by pliptor** [comment]
Okay, I figured out a way for me to test AVX with the help our our friend Travis.

[avx.patch.txt](https://github.com/tiny-dnn/tiny-dnn/files/846859/avx.patch.txt)

[results](https://travis-ci.org/pliptor/tiny-dnn-lab/builds/211631366)

If someone could verify if the avx sse test code is proper, it would be helpful.
Basically if the code is right, it is saying AVX calculations are twice the value of what SSE computes. It seems one bit shift off relative to each other.

AVX is computing almost four times of what CPU does. Maybe 2 bits shifts off. The trick is to test the lowest level functions and then go up. I think many places where we have 10-5 as passing criteria needs to be revised. At this low level the differences should be really tight.

**Comment by pliptor** [comment]
BTW, the differences above were for Travis/Linux. It seems Travis/macos has a wider discrepancy. It's in the [results](https://travis-ci.org/pliptor/tiny-dnn-lab/builds/211631366) as well.

**Comment by beru** [comment]
> Basically if the code is right, it is saying AVX calculations are twice the value of what SSE computes. It seems one bit shift off relative to each other.
> AVX is computing almost four times of what CPU does. Maybe 2 bits shifts off. The trick is to test the lowest level functions and then go up.

I'm not sure what `one bit shift off` and `2 bits shifts offf` mean in here.

`resemble` method accumulates the values holded by its type.
So according to your test code, `AVX calculations are twice the value of what SSE computes` is correct behavior.

**Comment by beru** [comment]
> AVX is computing almost four times of what CPU does.

It depends on the type you use. (Basically) SSE instructions use 128-bit XMM registers and AVX/AVX2 instructions use 256-bit YMM registers.

 &nbsp; | float | double |
-- | ---- | ----
sse | 4 | 2 |
avx | 8 | 4 |

**Comment by pliptor** [comment]
@beru Thank you very much for the detailed explanation. What I meant by `one bit shift off` is that the 2x difference seemed to be a bit shift (<<)  in the mantissa between AVX and SSE. But now you explained "resemble" accumulates the values within the XMM and YMM registers. I was then using the wrong method to read the results. So somehow the way I'm loading the registers duplicates the values within the registers. I think then if we keep bisecting we can try to find where CPU and AVX routine starts disagreeing (if they indeed do).

**Comment by pliptor** [comment]
@beru Thank you again for your help checking my earlier verification code. I fixed it in my own 
[branch](https://github.com/pliptor/tiny-dnn-lab/pull/2). Travis [results](https://travis-ci.org/pliptor/tiny-dnn-lab/builds/211961831).

For these particular tests SSE/AVX/CPU all match using EXPECT_EQ() at 10^-30 for float and 10^-100 for double final values (Linux tests).

**Comment by pliptor** [comment]
@beru Here's one good bisection point to start. Forward propagation stage of a single layer of identity activation. Do you think the differences are within reasonable ranges?

https://travis-ci.org/pliptor/tiny-dnn-lab/jobs/212175421

**Comment by pliptor** [comment]
This version compares the differences in error vector magnitude (EVM), a relative difference measure, and the output is expressed in parts per billion (PPM). My verification code needs some debugging as gcc and clang outputs don't agree.

https://travis-ci.org/pliptor/tiny-dnn-lab/jobs/212332784

**Comment by beru** [comment]
@pliptor Many computational operations aren't associative so I don't pursue to align calculation results.
Now I suppose the network sudden death (accuracy turns to 1000/10000) is due to exploding gradient problem. Since it happens with the internal backend too, I'm not sure if the problem is inherent inside the avx backend.

**Comment by pliptor** [comment]
@beru Thank you for the explanation.  [This paper](https://arxiv.org/pdf/1412.7024.pdf) and others support that neural network training is very forgiving with calculation precision so I agree with you that avx backend itself is then probably okay.

**Comment by beru** [comment]
@pliptor It took some time and I've found the way to prevent the problem from happening.
(The problem that accuracy suddenly drops to 1000/10000 around the peak of training with CIFAR-10 example train program)

In the train program, `cross_entropy` loss function is used.
https://github.com/tiny-dnn/tiny-dnn/blob/b89239c024a6f4d6e79f9f4108c30d483120691b/examples/cifar10/train.cpp#L111

And in `cross_entropy::df` method, the division is carried out but denominator can be zero.
https://github.com/tiny-dnn/tiny-dnn/blob/b89239c024a6f4d6e79f9f4108c30d483120691b/tiny_dnn/lossfunctions/loss_function.h#L123

A possible workaround is to add an epsilon value to the denominator.

**Comment by pliptor** [comment]
@beru Very nice finding.

**Comment by bhack** [comment]
And https://github.com/tiny-dnn/tiny-dnn/issues/410?

**Comment by pliptor** [comment]
@bhack It's most likely related by the same issue. cross_entropy_multiclass probably suffers as well.

**Comment by beru** [comment]
@bhack I think #410 is related too. https://github.com/tiny-dnn/tiny-dnn/issues/570#issuecomment-281405231

Now I hope I'll be able to reach > 80% accuracy with CIFAR-10 dataset...

**Comment by pliptor** [comment]
@bhack Did you have any luck? tiny-dnn does pretty well in MNIST. This is a funny quote in [tensorflow](https://www.tensorflow.org/get_started/mnist/pros) "Getting 92% accuracy on MNIST is bad. It's almost embarrassingly bad." I don't know what's considered bad for CIFAR-10 but I hope the current example can be considerably improved. I wonder what's the best people get without data shuffling and data augmentation techniques.

**Comment by bhack** [comment]
@pliptor http://rodrigob.github.io/are_we_there_yet/build/classification_datasets_results.html#43494641522d3130

**Comment by pliptor** [comment]
@bhack Yes, the reddit article is trying to parse Rodrigo's list, which is referenced within.

**Comment by beru** [comment]
@pliptor The short answer is "**No**". I've tried to improve accuracy by tweaking network model of example train of CIFAR-10 but I still haven't achieved considerable gain with training accuracy.

I tested several things such as changing type of activation functions to use (relu to elu), the number of in/out channels of conv layers (increasing the number of channels), tested batch normalization layer, etc...

Some of the findings. (though nothing new here I guess)

1. The network model of [ConvNetJS CIFAR-10 demo](http://cs.stanford.edu/people/karpathy/convnetjs/demo/cifar10.html) is more compact compared to tiny-dnn CIFAR-10 train example.
It takes around 90 seconds with tiny-dnn example network but 30 seconds with ConvNetJS demo model using my pc.

2. Lowering the learning rate **during** a training improves accuracy. Since it takes a lot of time to train NN, I can't help but to start with a high learning rate. This may lower accuracy but waiting hours is hard for me. My conclusion is, I need to find a way to train much faster.

3. Checked performance of GPU with tensorflow tutorial program
https://www.tensorflow.org/tutorials/deep_cnn
With my GTX 1070 card, training speed is like `(2160.0 examples/sec; 0.060 sec/batch)`
If you opened `tensorflow/models/tutorials/image/cifar10/cifar10_train.py`, it says

```
System        | Step Time (sec/batch)  |     Accuracy
------------------------------------------------------------------
1 Tesla K20m  | 0.35-0.60              | ~86% at 60K steps  (5 hours)
1 Tesla K40m  | 0.25-0.35              | ~86% at 100K steps (4 hours)
```

Since it takes around 4 to 5 hrs with somewhat old (few gens) GPUs.
I guess it's not that easy to achieve similar accuracy with CPU within 5 hrs then.

**Comment by pliptor** [comment]
@beru Wow you did quite a lot of experimenting. Thank you very much for so much interesting data. I hope it was fun for you. It then looks like checking or tracing accuracy is very difficult. Maybe when the field evolves a bit more we will have references to check against at some unit level.

**Comment by beru** [comment]
@pliptor The problem for me is I don't have solid foundation/understanding in the field. Very slow training reminds me of using offline 3D rendering software in the old days. Finding splendid sceneries within a procedural world seems possible as it holds quite a lot of possibilities/combination, but I need to enhance my abilities to pursue them.

An expert says `CIFAR-10` is a solved problem just like `MNIST` ([this interview](http://blog.kaggle.com/2014/12/22/convolutional-nets-and-cifar-10-an-interview-with-yan-lecun/)), so I feel at ease with the situation that I'm far behind.

**Comment by edgarriba** [comment]
>It takes around 90 seconds with tiny-dnn example network but 30 seconds with ConvNetJS demo model using my pc.

@beru what do you think is the difference in terms of implementation?

**Comment by Randl** [comment]
Well I managed to run a model which is claimed to achieve 90+% accuracy on CIFAR10, but it takes tons of time to train (~1 hour per epoch for 4 x E7-4860 v2 @ 2.60GHz, which means it'll take weeks to train). We need something relatively slim.

I try to look further on it.

**Comment by Randl** [comment]
Seems like ResNet is the best solution here)

**Comment by beru** [comment]
> what do you think is the difference in terms of implementation?
@edgarriba I don't get that question. I just tested the model on [this page](http://cs.stanford.edu/people/karpathy/convnetjs/demo/cifar10.html) with `tiny-dnn`.

**Comment by pliptor** [comment]
@beru maybe the question is why ConvNetJS network is 3 times faster to train than the one defined in tiny-dnn.
They are similar with respect to having 3 convolutional layers of size 5x5. But:

|Layer |ConvNetJS | tiny-dnn| 
|---|---|---|
|1 | 16 feature maps | 32 feature maps |
|2 | 20 feature maps| 64 feature maps |
|3| 20 feature maps| 64 feature maps |

Also, tiniy-dnn as an additional fully connected layer before the last layer.

**Comment by edgarriba** [comment]
@pliptor that's the question :smile:

**Comment by pliptor** [comment]
@edgarriba Cool. ConvNetJS does also a bit of data augmentation. That's another thing that shifts complexity away from the network itself for the accuracy/complexity trade-off.

**Comment by beru** [comment]
@pliptor The difference of the models is very clear so I wonder why edgar needed to ask. But thank you for the clarification.

I don't think 80% accuracy is possible with CnvNetJS CIFAR-10 model.
I tested it a few days ago and I could get was around 74%.
I also found [this post](https://groups.google.com/forum/#!msg/convnetjs/OwcXjkwECbg/s0QGqtMeFWwJ) .

**Comment by edgarriba** [comment]
@beru not sure if I'm missing something, just curious why convnetjs takes 30s and tiny-dnn 90s.

**Comment by pliptor** [comment]
@beru Nice. Someone should launch a competition for the best scheme that maximizes 
`
accuracy *(1/(training complexity)).
`
for CIFAR-10 dataset. :)

**Comment by beru** [comment]
> Well I managed to run a model which is claimed to achieve 90+% accuracy on CIFAR10, but it takes tons of time to train (~1 hour per epoch for 4 x E7-4860 v2 @ 2.60GHz, which means it'll take weeks to train). We need something relatively slim.

@Randl wow, If it takes weeks to train with such a pretty good spec computer, it's almost impossible to use that model with my pc.

**Comment by beru** [comment]
> not sure if I'm missing something, just curious why convnetjs takes 30s and tiny-dnn 90s.

@edgarriba Strictly speaking, I didn't compare convnetjs and tiny-dnn. I used a similar model written on ConvNetJS CIFAR-10 demo page with tiny-dnn and it takes about 30 secs per epoch with my pc. Why it only takes 30secs per epoch is, as I see it, because it uses fewer channels so it spends less time on mostly conv layers.

**Comment by beru** [comment]
> Someone should launch a competition for the best scheme that maximizes
accuracy *(1/(training complexity)).
for CIFAR-10 dataset. :)

@pliptor Considering the exponential relationship between accuracy and complexity of network models, it can be difficult to rank its value as it is a subjective matter. So having a graph plot would be good, or so I thought... (not my task.. not my task.. somebody in the field might have done it already)

**Comment by pliptor** [comment]
@beru Complexity is a bit more difficult to define but I think accuracy is well defined. Accuracy that I have in mind is the number of correct classification of images in the test set. 

@Randl has a nice computer. We could have a competition with the following metric
```
metric = accuracy^n * (1/(time to train in @Randl 's computer))^m.  
```
Typical commodity resources and total time of training is what matters for us, casual users, anyway. I wonder what are good `n` and `m` values. Dropping accuracy from 100% to 50% by reducing training time by 50% doesn't look like a good deal. Yeah, somebody must have looked at this metric already.

**Comment by vekkuli** [comment]
I've been playing around with tiny-dnn and one of my issues was this sudden network collapse.

@beru 
> A possible workaround is to add an epsilon value to the denominator.

I made this modification:
`d[i] = (y[i] - t[i]) / (y[i] * (float_t(1) - y[i]) + std::numeric_limits<float>::epsilon());`
and it seems to help.

The setup is as follows:
```
net     << convolutional_layer(32, 32, 5, 1, n_fmaps, padding::same)		// C1
	<< max_pooling_layer(32, 32, n_fmaps, 2)							// P2
	<< relu_layer(16, 16, n_fmaps)										// activation
	<< convolutional_layer(16, 16, 5, n_fmaps, n_fmaps, padding::same)  // C3
	<< max_pooling_layer(16, 16, n_fmaps, 2, backend_type)				// P4
	<< relu_layer(8, 8, n_fmaps)										// activation
	<< convolutional_layer(8, 8, 5, n_fmaps, n_fmaps2, padding::same)   // C5
	<< max_pooling_layer(8, 8, n_fmaps2, 2)								// P6
	<< relu_layer(4, 4, n_fmaps2)										// activation
	<< fc(4 * 4 * n_fmaps2, n_fc)										// FC7
	<< relu_layer(n_fc)													// activation
	<< fc(n_fc, 3) << softmax(3);										// FC10
adagrad opt;```

**Comment by radioman** [comment]
..instead of simplifying things its bloating into uncomprehendable complexity, it seems AI is going in the wrong way... 'training' for weeks for simple enough puzzle looks pretty insane ;/ its far from 'solved'

**Comment by pliptor** [comment]
@radioman It sure makes us want to revisit simpler things that can get the job done. I've been running different trade-offs with other techniques as well. On the other hand, insanity fuels good research too, so a bit welcome too.

**Comment by beru** [comment]
> I made this modification:
d[i] = (y[i] - t[i]) / (y[i] * (float_t(1) - y[i]) + std::numeric_limits<float>::epsilon());
and it seems to help.

@vekkuli Yes, I did the same thing and there can be an another way to prevent division by zero.
```cpp
d[i] = (y[i] - t[i]) / std::max<float_t>(std::numeric_limits<float_t>::epsilon(), y[i] * (float_t(1) - y[i]));
```

And I wonder if `cross_entropy::f` method should be taken care of. Because it uses `std::log` function so passing zero should be avoided. But so far, I still haven't faced with this problem.
